File: php/src/VersionedEncryptor.php
<?php

declare(strict_types=1);

namespace SecureKit;

/**
 * Wraps SymmetricEncryptor with key-ID-tagged ciphertexts, so keys can be
 * rotated without losing the ability to decrypt data encrypted under a
 * previous key. Wire format: 1-byte key ID followed by whatever
 * SymmetricEncryptor::encrypt() produces. Supports up to 256 concurrently-
 * known key versions.
 */
final class VersionedEncryptor
{
    private SymmetricEncryptor $inner;
    /** @var array<int, string> */
    private array $keys;
    private int $currentId;

    /**
     * @param array<int, string> $keys keyId (0-255) => raw 32-byte key
     * @throws \InvalidArgumentException if $currentKeyId is not in $keys
     */
    public function __construct(array $keys, int $currentKeyId)
    {
        if (!array_key_exists($currentKeyId, $keys)) {
            throw new \InvalidArgumentException('securekit: current key ID not found in key set');
        }
        $this->inner = new SymmetricEncryptor();
        $this->keys = $keys;
        $this->currentId = $currentKeyId;
    }

    /** Registers a new key version without changing which key is current. */
    public function addKey(int $id, string $key): void
    {
        $this->keys[$id] = $key;
    }

    /**
     * Switches which registered key new encrypt() calls use.
     * @throws \InvalidArgumentException if $id was never registered
     */
    public function setCurrentKeyId(int $id): void
    {
        if (!array_key_exists($id, $this->keys)) {
            throw new \InvalidArgumentException('securekit: key ID not found in key set');
        }
        $this->currentId = $id;
    }

    public function encrypt(string $plaintext, string $aad = ''): string
    {
        $blob = $this->inner->encrypt($this->keys[$this->currentId], $plaintext, $aad);
        return chr($this->currentId) . $blob;
    }

    /**
     * @throws \RuntimeException with a generic message on any failure,
     * including an unrecognized key ID, to avoid leaking which key IDs
     * are valid.
     */
    public function decrypt(string $blob, string $aad = ''): string
    {
        if (strlen($blob) < 1) {
            throw new \RuntimeException('securekit: decryption failed');
        }
        $id = ord($blob[0]);
        if (!array_key_exists($id, $this->keys)) {
            throw new \RuntimeException('securekit: decryption failed');
        }
        return $this->inner->decrypt($this->keys[$id], substr($blob, 1), $aad);
    }
}

File: python/securekit/versioned_encryptor.py
"""Wraps SymmetricEncryptor with key-ID-tagged ciphertexts, so keys can be
rotated without losing the ability to decrypt data encrypted under a
previous key."""

from .symmetric_encryptor import SymmetricEncryptor, DecryptionError


class VersionedEncryptor:
    """Wire format: 1-byte key ID followed by whatever
    SymmetricEncryptor.encrypt() produces. Supports up to 256 concurrently-
    known key versions."""

    def __init__(self, keys: dict[int, bytes], current_key_id: int) -> None:
        if current_key_id not in keys:
            raise ValueError("securekit: current key ID not found in key set")
        self._inner = SymmetricEncryptor()
        self._keys = dict(keys)
        self._current_id = current_key_id

    def add_key(self, key_id: int, key: bytes) -> None:
        """Registers a new key version without changing which key is current."""
        self._keys[key_id] = key

    def set_current_key_id(self, key_id: int) -> None:
        if key_id not in self._keys:
            raise ValueError("securekit: key ID not found in key set")
        self._current_id = key_id

    def encrypt(self, plaintext: bytes, aad: bytes | None = None) -> bytes:
        blob = self._inner.encrypt(self._keys[self._current_id], plaintext, aad)
        return bytes([self._current_id]) + blob

    def decrypt(self, blob: bytes, aad: bytes | None = None) -> bytes:
        """Raises the generic DecryptionError on any failure, including an
        unrecognized key ID, to avoid leaking which key IDs are valid."""
        if len(blob) < 1:
            raise DecryptionError()
        key_id = blob[0]
        if key_id not in self._keys:
            raise DecryptionError()
        return self._inner.decrypt(self._keys[key_id], blob[1:], aad)
